node/src/components/storage/lmdb_block_height_store.rs

```diff
@@ -0,0 +1,85 @@
+use std::{
+    fmt::Debug,
+    path::Path,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use lmdb::{
+    self, Cursor, Database, DatabaseFlags, Environment, EnvironmentFlags, Transaction, WriteFlags,
+};
+use serde::{Deserialize, Serialize};
+use tracing::info;
+
+use super::{BlockHeightStore, Result};
+
+/// LMDB version of a store.
+#[derive(Debug)]
+pub(super) struct LmdbBlockHeightStore {
+    env: Environment,
+    db: Database,
+    highest: AtomicU64,
+}
+
+impl LmdbBlockHeightStore {
+    pub(crate) fn new<P: AsRef<Path>>(db_path: P, max_size: usize) -> Result<Self> {
+        let env = Environment::new()
+            .set_flags(EnvironmentFlags::NO_SUB_DIR)
+            .set_map_size(max_size)
+            .open(db_path.as_ref())?;
+        let db = env.create_db(None, DatabaseFlags::INTEGER_KEY)?;
+
+        // Get the last key, which will represent the largest block height, since the LMDB is sorted
+        // by integer key increasing.
+        let mut max_height_bytes = [0; 8];
+        let txn = env.begin_ro_txn().expect("should create ro txn");
+        {
+            let mut cursor = txn.open_ro_cursor(db).expect("should create ro cursor");
+            for (height_bytes, _value) in cursor.iter() {
+                max_height_bytes.copy_from_slice(height_bytes);
+            }
+        }
```

Comment:
We are iterating over _all_ blocks in the storage? No, we shouldn't be doing that :grimacing: We should have a separate namespace/table in the LMDB database that will store only one value – the highest block and it should be updated in one transaction, together with:
* insert block 
* insert height -> block hash
* update the highest block entry

any other solution is just temporary.

Reply:
I think what's in this PR is a simpler solution, and not unreasonable, at least for now.

Rather than hitting the DB every time a block is stored, and every time a request for the highest block is made, this is instead just a call to an atomic u64.  The iteration over all entries in the block-height store is only done once per session - on startup - to set the initial value of the atomic u64.  Even with millions of entries, it should be pretty fast, since we're not actually deserializing any of the heights or hashes except the last one.

If the startup performance does become an issue, we can look to store the heights in reverse order to allow us to only need to read the first value on startup.

Reply:
>  since we're not actually deserializing any of the heights or hashes except the last one.

Are you sure that the cursor will return elements in the order in which they were inserted?

Reply:
No, it's not ordered by insertion order, but rather lexicographical order, or in the case of integer keys, numerical ascending order.

The included test which randomly generates heights should ensure this remains a valid approach.

Reply:
> The included test which randomly generates heights should ensure this remains a valid approach.

Ah, I forgot about that by now. Thanks.